File: generator.rb
def get_generator(batch_size = 10)
  items = (0..999).to_a
  Enumerator.new do |enum|
    i = 0
    while true
      offset = batch_size * i
      _items = items[offset, batch_size]
      if _items == nil || _items.length == 0
        break
      end

      enum.yield _items
      i +=1
    end
  end
end

g = get_generator
g.each do |items|
  p items
end

File: generator.js
function getGenerator(batchSize = 10) {
  return function* () {
    const items = [...Array(1000).keys()];
    let i = 0;
    while (true) {
      const offset = batchSize * i;
      const _items = items.slice(offset, offset + batchSize);
      if (_items.length === 0) {
        break;
      }
      i ++;
      yield _items;
    }
  }
}


const it = {};
it[Symbol.iterator] = getGenerator();

for (let items of it) {
  console.log(items);
}


